### ISSUE TITLE
MSBP File Format

### ISSUE THREAD

**Issue by kenshiisod** [kinnay/Nintendo-File-Formats#3]
Hi! Just pointing out an MSBP file can be used to reverse the bytes back into Nintendo's original bbcode-like syntax.

Download 3DS System Transfer (CARDBOARD, 00022A00) and open the MSBT/MSBP files.
Look at the indexes of the tag groups and tags in the MSBP file, you see:

| Tag group | Index | Hex |
| --- | --- | --- |
| System | 0 | 0x00 |
| UserName | 1 | 0x01 |
| DayTime | 2 | 0x02 |
| SoftInfo | 3 | 0x03 |
| SettingInfo | 4 | 0x04 |
| NandInfo | 5 | 0x05 |
| PointInfo | 6 | 0x06 |
| DataInfo | 7 | 0x07 |
| ErrorInfo | 8 | 0x08 |
| Blink | 9 | 0x09 |
| ProgressInfo | 10 | 0x0a |
| MigrateInfo | 11 | 0x0b |

Those tag groups correspond to the first byte of the command (they're the possible [command types](https://github.com/kinnay/Nintendo-File-Formats/wiki/MSBT-File-Format#txt2-block) for the game).

The tags for each tag group, also known as subcommands, are also there. Looking at a few, based on what we know:

For UserName (tag group 0x01):

| Tag | Index | Hex |
| --- | --- | --- |
| MyName | 0 | 0x00 |
| TargetName | 1 | 0x01 |
| SenderName | 2 | 0x02 |
| ReceiverName | 3 | 0x03 |

Looking in cardboard_ctr.msbt we see these commands in bytes. We can convert to plaintext using the above info:
| Bytes | Plaintext |
| --- | --- |
| 0E00 0100 0000 0000 | `[UserName::MyName ]`
| 0E00 0100 0100 0000 | `[UserName::TargetName ]`
| 0E00 0100 0200 0000 | `[UserName::SenderName ]`
| 0E00 0100 0300 0000 | `[UserName::ReceiverName ]`

In the first line above 0100 = tag group, 0000 = tag, 0000 = 0 more utf8 bytes relating to this command
So we can convert bytes <-> plaintext using the MSBP file, although the parameters would probably need to be done manually.

The UserName ones are just variables, whereas most of the System ones take parameters.
For example: `[System::Ruby rt="ちい"]小[/System::Ruby]`
Parameters such as `rt` are listed under `TGP2`.

It looks like ACNH on the Switch has App.msbp which doesn't contain TAG, TGG, TGP, etc. so this isn't possible, which is weird.
Maybe Nintendo removed the strings / those sections since there's no reason for the game to use them, only the translation tool?

Hopefully some of this braindump helps, but I'm not sure if you already knew that?

Maybe worth mentioning the relationship somewhere in the wiki? If a game has an MSBP, conversion can be mostly automated.

**Comment by shane-tw** [comment]
Also, side-note, ms_tags.h (included in various 3ds system titles' romfs) mentions this:
```
#define MS_TAG_BEGIN_MARK       0x0E
#define MS_TAG_END_MARK         0x0F
```

We know 0x0E marks the beginning of a tag, I guess 0x0F marks the end of some custom ones? I haven't seen it used.
Just since the docs mention this:
> Character 0xF is a bit different. It always takes up exactly 6 bytes (including the 0xF character). Its purpose is unknown.

**Comment by kenshiisod** [comment]
Just adding a note that 0xF means *tag end marker* in the case of tags that follow the standard format. 

Since the docs mention:
> Character 0xF is a bit different. It always takes up exactly 6 bytes (including the 0xF character). Its purpose is unknown.

If you look at EUnl/SP_owl_Comment_Fossil.msbt in ACNH, there's an example of it being used.
Here's that in PO format: [EUnl/SP_owl_Comment_Fossil.po](https://gist.github.com/kenshiisod/9431180d4fb865ac855a4e680112fca1#file-sp_owl_comment_fossil-po-L766)
ZIP attached containing the original MSBT and the same PO: [EUnl.zip](https://github.com/kinnay/Nintendo-File-Formats/files/6300434/EUnl.zip)

**Comment by kinnay** [comment]
Hi, sorry for the late reply. Thanks for your research. I never realized that tag groups in the MSBP file define the message control tags, but it all makes sense now.

I updated the wiki page about MSBT files. Let me know if anything is still missing, or incorrect. I don't think I want to list the game-specific tags for each game but I added a note that explains that the tags are defined in the MSBP file.

Thanks again.
